File: core/tests/test_tree.cpp
#include <gtest/gtest.h>
#include "tree.h"
#include <string>

class TestObserver : public ITreeObserver<std::string> {
public:
    bool nodeAdded = false;
    bool nodeRemoved = false;
    bool nodeChanged = false;
    
    void onNodeAdded(ITreeNode<std::string>* node) override {
        nodeAdded = true;
    }
    
    void onNodeRemoved(ITreeNode<std::string>* node) override {
        nodeRemoved = true;
    }
    
    void onNodeChanged(ITreeNode<std::string>* node) override {
        nodeChanged = true;
    }
    
    void reset() {
        nodeAdded = false;
        nodeRemoved = false;
        nodeChanged = false;
    }
};

class TestVisitor : public ITreeVisitor<std::string> {
public:
    std::vector<std::string> visitedNodes;
    
    void visitEnter(ITreeNode<std::string>* node) override {
        visitedNodes.push_back("enter:" + node->getName());
    }
    
    void visitLeave(ITreeNode<std::string>* node) override {
        visitedNodes.push_back("leave:" + node->getName());
    }
};

// PathUtils测试
TEST(PathUtilsTest, BasicOperations) {
    EXPECT_EQ(PathUtils::getParentPath("/a/b/c"), "/a/b");
    EXPECT_EQ(PathUtils::getParentPath("/"), "/");
    EXPECT_EQ(PathUtils::getParentPath(""), "/");
    
    EXPECT_EQ(PathUtils::getLastSegment("/a/b/c"), "c");
    EXPECT_EQ(PathUtils::getLastSegment("/"), "");
    
    EXPECT_EQ(PathUtils::joinPaths("/a/b", "c"), "/a/b/c");
    EXPECT_EQ(PathUtils::joinPaths("/a/b/", "/c"), "/a/b/c");
    
    EXPECT_TRUE(PathUtils::isAbsPath("/a/b"));
    EXPECT_FALSE(PathUtils::isAbsPath("a/b"));
    
    auto parts = PathUtils::parse("/a/b/c");
    EXPECT_EQ(parts.size(), 3);
    EXPECT_EQ(parts[0], "a");
    EXPECT_EQ(parts[1], "b");
    EXPECT_EQ(parts[2], "c");
}

TEST(PathUtilsTest, NormalizeAndRelativePath) {
    EXPECT_EQ(PathUtils::normalize("/a/b/../c"), "/a/c");
    EXPECT_EQ(PathUtils::normalize("/a/./b/./c"), "/a/b/c");
    EXPECT_EQ(PathUtils::normalize("/a/b/../../c"), "/c");
    
    EXPECT_EQ(PathUtils::parseRelativePath("../b", "/a/c"), "/a/b");
    EXPECT_EQ(PathUtils::parseRelativePath("./b", "/a/c"), "/a/c/b");
    EXPECT_EQ(PathUtils::parseRelativePath("b", "/a/c"), "/a/c/b");
    
    EXPECT_EQ(PathUtils::getRelativePath("/a/b/c", "/a"), "b/c");
    EXPECT_EQ(PathUtils::getRelativePath("/a/d", "/a/b/c"), "../../d");
}

// Tree基本操作测试
class TreeTest : public ::testing::Test {
protected:
    Tree<std::string>* tree;
    TestObserver* observer;
    
    void SetUp() override {
        tree = new Tree<std::string>();
        observer = new TestObserver();
        tree->addObserver(observer);
    }
    
    void TearDown() override {
        delete tree;
        delete observer;
    }
};

TEST_F(TreeTest, BasicOperations) {
    // 测试初始状态
    EXPECT_EQ(tree->getName(), "");
    EXPECT_EQ(tree->getParent(), nullptr);
    EXPECT_TRUE(tree->getChildren().empty());
    
    // 添加子节点
    auto child1 = new TreeNode<std::string>("child1");
    child1->setName("child1");
    tree->addChild(child1);
    
    EXPECT_TRUE(observer->nodeAdded);
    EXPECT_EQ(tree->getChildren().size(), 1);
    EXPECT_EQ(child1->getParent(), tree);
    
    // 移除子节点
    tree->removeChild(child1);
    EXPECT_TRUE(observer->nodeRemoved);
    EXPECT_TRUE(tree->getChildren().empty());
}

TEST_F(TreeTest, PathOperations) {
    auto child1 = new TreeNode<std::string>("child1");
    child1->setName("child1");
    auto child2 = new TreeNode<std::string>("child2");
    child2->setName("child2");
    
    tree->addChild(child1);
    child1->addChild(child2);
    
    EXPECT_EQ(child2->getPath(), "/child1/child2");
    EXPECT_EQ(tree->findByPath("/child1/child2"), child2);
    EXPECT_EQ(tree->findByPath("/nonexistent"), nullptr);
    
    EXPECT_EQ(child1->getRelativePath(child2), "child2");
    EXPECT_EQ(child2->getRelativePath(child1), "..");
}

TEST_F(TreeTest, TraversalAndSearch) {
    auto child1 = new TreeNode<std::string>("child1");
    child1->setName("child1");
    auto child2 = new TreeNode<std::string>("child2");
    child2->setName("child2");
    
    tree->addChild(child1);
    child1->addChild(child2);
    
    TestVisitor visitor;
    tree->traverse(&visitor);
    
    EXPECT_EQ(visitor.visitedNodes.size(), 6); // 3 nodes * 2 (enter/leave)
    EXPECT_EQ(visitor.visitedNodes[0], "enter:");
    EXPECT_EQ(visitor.visitedNodes[1], "enter:child1");
    EXPECT_EQ(visitor.visitedNodes[2], "enter:child2");
    
    // 测试查找
    auto found = tree->find([](ITreeNode<std::string>* node) {
        return node->getName() == "child2";
    });
    EXPECT_EQ(found, child2);
    
    found = tree->find([](ITreeNode<std::string>* node) {
        return node->getName() == "nonexistent";
    });
    EXPECT_EQ(found, nullptr);
}

TEST_F(TreeTest, DataOperations) {
    auto child = new TreeNode<std::string>("data1");
    child->setName("child");
    
    EXPECT_TRUE(child->getData().has_value());
    EXPECT_EQ(child->getData().value(), "data1");
    
    tree->addChild(child);
    EXPECT_FALSE(tree->getData().has_value());
}

int main(int argc, char **argv) {
    testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: pseudo/test_tree.py
import unittest
from typing import List, Optional

from mytree import *

class TestTree(unittest.TestCase):
    def setUp(self):
        # 创建测试用的树结构
        self.tree = Tree[str]()
        self.root_node = TreeNode[str]("root data")
        self.root_node.set_name("root")
        self.tree.add_child(self.root_node)
        
        self.child1 = TreeNode[str]("child1 data")
        self.child1.set_name("child1")
        self.root_node.add_child(self.child1)
        
        self.child2 = TreeNode[str]("child2 data")
        self.child2.set_name("child2")
        self.root_node.add_child(self.child2)
        
        self.grandchild = TreeNode[str]("grandchild data")
        self.grandchild.set_name("grandchild")
        self.child1.add_child(self.grandchild)

    def test_basic_operations(self):
        # 测试基本的树操作
        self.assertEqual(self.root_node.get_name(), "root")
        self.assertEqual(len(self.root_node.get_children()), 2)
        self.assertEqual(self.child1.get_parent(), self.root_node)
        self.assertEqual(self.child1.get_data(), "child1 data")

    def test_path_operations(self):
        # 测试路径相关操作
        self.assertEqual(self.grandchild.get_path(), "/root/child1/grandchild")
        self.assertEqual(self.child2.get_path(), "/root/child2")
        
        # 测试相对路径
        relative_path = self.child1.get_relative_path(self.child2)
        self.assertEqual(relative_path, "../child2")
        
        relative_path = self.grandchild.get_relative_path(self.child2)
        self.assertEqual(relative_path, "../../child2")

    def test_find_operations(self):
        # 测试查找操作
        found_node = self.tree.find_by_path("/root/child1/grandchild")
        self.assertEqual(found_node, self.grandchild)
        
        # 测试条件查找
        found_node = self.tree.find(lambda node: node.get_data() == "child2 data")
        self.assertEqual(found_node, self.child2)

    def test_path_utils(self):
        # 测试路径工具类
        self.assertEqual(PathUtils.get_parent_path("/root/child1"), "/root")
        self.assertEqual(PathUtils.get_last_segment("/root/child1"), "child1")
        self.assertEqual(PathUtils.normalize("/root/./child1/../child2"), "/root/child2")
        self.assertEqual(PathUtils.join_paths("/root", "child1"), "/root/child1")

class TestTreeObserver(ITreeObserver[str]):
    def __init__(self):
        self.added_nodes: List[ITreeNode[str]] = []
        self.removed_nodes: List[ITreeNode[str]] = []
        self.changed_nodes: List[ITreeNode[str]] = []

    def on_node_added(self, node: ITreeNode[str]) -> None:
        self.added_nodes.append(node)

    def on_node_removed(self, node: ITreeNode[str]) -> None:
        self.removed_nodes.append(node)

    def on_node_changed(self, node: ITreeNode[str]) -> None:
        self.changed_nodes.append(node)

class TestTreeVisitor(ITreeVisitor[str]):
    def __init__(self):
        self.visited_nodes: List[str] = []

    def visit_enter(self, node: ITreeNode[str]) -> None:
        self.visited_nodes.append(f"enter:{node.get_name()}")

    def visit_leave(self, node: ITreeNode[str]) -> None:
        self.visited_nodes.append(f"leave:{node.get_name()}")

class TestObserverAndVisitor(unittest.TestCase):
    def setUp(self):
        self.tree = Tree[str]()
        self.root_node = TreeNode[str]("root")
        self.root_node.set_name("root")
        self.observer = TestTreeObserver()
        self.root_node.observers.append(self.observer)

    def test_observer(self):
        # 测试观察者模式
        child = TreeNode[str]("child")
        child.set_name("child")
        self.root_node.add_child(child)
        self.assertEqual(len(self.observer.added_nodes), 1)
        
        self.root_node.remove_child(child)
        self.assertEqual(len(self.observer.removed_nodes), 1)

    def test_visitor(self):
        # 测试访问者模式
        child = TreeNode[str]("child")
        child.set_name("child")
        self.root_node.add_child(child)
        
        visitor = TestTreeVisitor()
        self.root_node.traverse(visitor)
        
        expected_visits = [
            "enter:root",
            "enter:child",
            "leave:child",
            "leave:root"
        ]
        self.assertEqual(visitor.visited_nodes, expected_visits)

if __name__ == '__main__':
    unittest.main()